docs/man/mutool.1
.TH "MUTOOL" "1" "June 3, 2026"
.\" Please adjust this date whenever revising the manpage.
.\" no hyphenation
.nh
.\" adjust left
.ad l
.
.SH NAME
mutool \- all purpose tool for dealing with PDF files
.
.SH SYNOPSIS
.SY mutool
.I sub-command
.RI [ options ]
.YS
.
.SH DESCRIPTION
mutool is a tool based on MuPDF for dealing with document files in various
manners. There are several sub commands available, as described below.
.
.SH VERSION
.SY mutool
.B \-v
.YS
.PP
Shows the MuPDF version used to build mutool.
.
.SH AUDIT
.SY mutool
.B audit
.RI [ options ]
.I file1.pdf
.RI [ file2.pdf\~ .\|.\|.]
.YS
.PP
Outputs a report of operator and space usage for each given PDF file.
.TP
.BI \-o\~ filename
The output filename for the report.
.
.SH BAKE
.SY mutool
.B bake
.RI [ options ]
.I input.pdf
.RI [ output.pdf ]
.YS
.PP
Normally annotations and form field widgets are separate content streams
that get drawn on top of the page contents. This command bakes annotations
and/or widgets into the page contents, making them inseparable.
After baking in annotations and widgets they are just visual representaitons
of what the original annotation/widget looked like, their properties can no
longer be edited since they no longer exist as proper annotation/widgets.
.TP
.B \-A
Do not bake in annotations.
.TP
.B \-F
Do not bake in form field widgets.
.TP
.BI \-O\~ options
A comma-separated list of PDF options. See
.B PDF OUTPUT OPTIONS
below.
.
.SH BARCODE
.SY mutool
.B barcode -d
.RI [ options ]
.I file1.pdf
.RI [ pages1 ]
.RI [ file2.pdf ]
.RI [ pages2 ]
.YS
.PP
Decodes a barcode from an image or a document.
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.BI \-o\~ output
The output filename.
If no output is specified, any text from decoded barcodes will go to stdout.
.TP
.BI \-r\~ angle
Hwo much to rotate the input pages in degrees (0-360), before
trying to decode any barcodes.
.TP
.I pages
Comma-separated list of page numbers and ranges (for example: 1,5,10-15,20-N),
where the character N denotes the last page. If no pages are specified, then all
pages will be processed.
.
.SY mutool
.B barcode -c
.RI [ options ]
.I text
.YS
.PP
Creates a new barcode containing text argument and formatted
according to options specified.
.TP
.BI \-o\~ output
The output format is inferred from the output filename, either PNG or PDF.
If no output is specified, the default is
.B out.png
.TP
.BI \-F\~ format
The desired output barcode format:
.in +4n
.B aztec
.br
.B codabar
.br
.B code39
.br
.B code93
.br
.B code128
.br
.B databar
.br
.B databarexpanded
.br
.B datamatrix
.br
.B ean8
.br
.B ean13
.br
.B itf
.br
.B maxicode
.br
.B pdf417
.br
.B qrcode
.br
.B upca
.br
.B upce
.br
.B microqrcode
.br
.B rmqrcode
.br
.B dxfilmedge
.br
.B databarlimited
.br
.TP
.BI \-s\~ size
Set size in pixels for the output barcode. If not specified the
smallest size that can be decoded is chosen.
.TP
.B \-q
Add quiet zones around the barcode. This puts an empty margin
around created barcodes.
.TP
.B \-t
Add human-readable text, when available. Some barcodes, e.g.
EAN-13, may have the barcode contents printed in human-readable
text next to the barcode, which is enabled by this flag.
.TP
.BI \-t\~ level
Set error correction level (0-8). Some barcodes, e.g. QR-codes,
support several levels of error correction, which can be
customized through this option.
.TP
.I text
The text to be encoded into a barcode.
.
.SH CLEAN
.SY mutool
.B clean
.RI [ options ]
.I input.pdf
.RI [ output.pdf ]
.RI [ pages\~ .\|.\|.]
.YS
.PP
The clean command pretty prints and rewrites the syntax of a PDF file.
It can be used to repair broken files, expand compressed streams, filter
out a range of pages, etc.
.PP
If no output file is specified, it will write the cleaned PDF to "out.pdf"
in the current directory.
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.B \-g
Garbage collect objects that have no references from other objects.
.TP
.B \-gg
In addition to
.BR \-g ,
also renumber all objects and compact the cross reference table.
.TP
.B \-ggg
In addition to
.BR \-gg ,
also merge and reuse duplicate objects.
.TP
.B \-gggg
In addition to
.BR \-ggg ,
also merge and reuse duplicate streams.
.TP
.B \-l
.B No longer supported!
Linearize output. Used to create a "Web Optimized" output file.
.TP
.B \-D
Save file without encryption.
.TP
.BI \-E\~ method
Save file with desired encryption method, one of:
.BR rc4-40 , rc40-128 , aes-128 ,
or
.BR aes-256 .
.TP
.BI \-O\~ password
Owner password used when encrypting output file. This is required to edit the
output file.
.TP
.BI \-U\~ password
User password used when encrypting output file. This is required to view the
output file.
.TP
.BI \-P\~ permissions
Permissions used when encrypting output file. These are the permissions that
apply when the output file is viewed.
.TP
.B \-a
ASCII Hex encode binary streams. Use in conjunction with
.B -d
and
.B -i
or
.B -f
to ensure that although the images and/or fonts are compressed, the resulting
file can still be viewed and edited with a text editor.
.TP
.B \-d
Decompress streams. This will make the output file larger, but provides
easy access for reading and editing the contents with a text editor.
.TP
.B \-z
Deflate uncompressed streams.
.br
If combined with
.BR -d ,
any decompressed streams will be recompressed.
.br
If combined with
.BR -a ,
the streams will also be hex encoded after compression.
.TP
.BI \-e\~ effort
Configure how much effort is spent compressing the output file.
.B 0
is default amount, otherwise a percentage of the maximum effort,
.B 1
for minimum and
.B 100
for maximum effort.
.TP
.B \-f
Toggle decompression of font streams. Use in conjunction with
.B -d
to leave fonts compressed.
.TP
.B \-i
Toggle decompression of image streams. Use in conjunction with
.B -d
to leave images compressed.
.TP
.B \-c
Pretty-print graphics commands in content streams.
.TP
.B \-s
Sanitize graphics commands in content streams, this means, for example, removing
invalid or redundant commands.
.TP
.B \-t
Use compact PDF object syntax. This saves space.
.TP
.B \-tt
Use indented PDF object syntax. This is more human-readable.
.TP
.B \-L
Print comments containing labels showing how each object can be reached from the
Root.
.TP
.B \-v
Transform all text into vector graphics commands.
.TP
.B \-A
Create appearance streams for annotations that are missing appearance streams.
.TP
.B \-AA
Recreate appearance streams for all annotations.
.TP
.B \-m
Preserve metadata.
.TP
.B \-S
Subset fonts if possible. (EXPERIMENTAL!)
.TP
.B \-Z
Use object streams cross reference streams for extra compressoin.
.TP
.BR \-\- ( color | gray | bitonal )\c
.BR \- (\c
.RB | lossy- | lossless- )\c
.BI image-subsample-method= method
Set the subsampling method
.RB ( average\~ or\~ bicubic )
for the
desired image types, for example color-lossy and bitonal-lossless.
.TP
.BR \-\- ( color | gray | bitonal )\c
.BR \- (\c
.RB | lossy- | lossless- )\c
.BI image-subsample-dpi= dpi\c
.RI [, dpi ]
Set the resolution at which to subsample.
.TP
.BR \-\- ( color | gray | bitonal )\c
.BR \- (\c
.RB | lossy- | lossless- )\c
.BI image-recompress-method= quality
Set the recompression quality to either of
.BR never ,
.BR same ,
.BR lossless ,
.BR jpeg ,
.BR j2k ,
.BR fax,
or
.BR jbig2 .
.TP
.BI \-\-recompress-images-when=\c
.RB ( smaller | always )
Recompress images when the rewritten images by the subsampling- or
recompression methods above are
.BR smaller\~ (default),
or
.B always
recompress them.
.TP
.BR \-\-structure =\c
.RB ( keep | drop )
Keep or drop the structure tree.
.TP
.I pages
Comma-separated list of page numbers and ranges (for example: 1,5,10-15,20-N),
where the character N denotes the last page. If no pages are specified, then all
pages will be included.
.
.SH CONVERT
.SY mutool
.B convert
.RI [ options ]
.I input.pdf
.RI [ pages\~ .\|.\|.]
.YS
.PP
The convert command is used to convert a file from one format to another.
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.BI \-o\~ output
The output format is inferred from the output filename.
Embed %d in the name to indicate the page number (for example: "page%d.png").
Printf modifiers are supported, for example "%03d".
If no output is specified, the output will go to stdout.
.TP
.BI \-F\~ format
Enforce a specific output format. Only necessary when outputting to stdout
since normally the output filename is used to infer the output format.
.TP
.BI \-O\~ options
Comma-separated list of format specific output options, see
.B PDF OUTPUT OPTIONS
and
.B OUTPUT OPTIONS
below.
.TP
.BI \-b\~ box
Use named page box
.RB ( MediaBox ,
.BR CropBox ,
.BR BleedBox ,
.BR TrimBox ,
or
.BR ArtBox ).
.TP
.BI \-A\~ bits
Specify how many bits of anti-aliasing to use. The default is 8.
.TP
.BI \-W\~ width
Page width in points for EPUB layout.
.TP
.BI \-H\~ height
Page height in points for EPUB layout.
.TP
.BI \-S\~ size
Font size in points for EPUB layout.
.TP
.BI \-U\~ filename
User CSS stylesheet for EPUB layout.
.B \-X
Disable document styles for EPUB layout.
.TP
.I pages
Comma-separated list of page numbers and ranges (for example: 1,5,10-15,20-N),
where the character N denotes the last page. If no pages are specified, then all
pages will be rendered.
.
.SH CREATE
.SY mutool
.B create
.RB [ \-o
.IR output.pdf ]
.RI [ options ]
.I page.txt
.RI [ page2.txt\~ .\|.\|.]
.YS
.PP
The create command creates a new PDF file with the contents created
from one or more input files containing graphics commands.
.TP
.BI \-o\~ output
If no output file is specified, it will write the created PDF to "out.pdf"
in the current directory.
.TP
.BI \-O\~ options
Comma-separated list of format specific output options, see
.B PDF OUTPUT OPTIONS
below.
.TP
.B page.txt
A page is created for each input file, with the contents of the file copied
into the content stream. Special comments in the input files are parsed to
define the page dimensions and font and image resources:
.RS
.TP
.B %%MediaBox LLX LLY URX URY
Set the page size (default A4).
.TP
.B %%Rotate Angle
Set the page rotation (default 90 degrees).
.TP
.B %%Font Name Filename Encoding
Create a simple font resource using the declared
.BR Name .
.B Filename
can be either a font file or one of the Base-14 font names
.RB ( Times-Roman\c
, etc).
Encoding is one of
.B Latin
(for ISO-8859-1),
.B Greek
(for ISO-8859-7) or
.B Cyrillic
(for KOI-8).
.TP
.B %%Font Name Language WMode Style Language
Create a CID font resource for CJK scripts using the declared
.BR Name .
.B Language
is one of
.BR zh-Hant ,
.BR zh-Hans ,
.BR ja ,
or
.BR ko .
.B Wmode
is one of
.BR H ,
or
.BR V .
.BR Style
is
.BR serif ,
or
.BR sans .
.TP
.B %%Image Im0 path/to/image.png
Create an image resource.
.TP
.B COMPLETE EXAMPLE
This adds an image, a triangle and some text to a page:
.in +4n
.sp
.EX
%%MediaBox 0 0 595 842
%%Font TmRm Times-Roman
%%Font Helv-C Helvetica Cyrillic
%%Font Helv-G Helvetica Greek
%%CJKFont Song zh-Hant
%%CJKFont Mincho ja
%%CJKFont Batang ko
%%Image I0 docs/examples/huntingofthesnark.png
.sp
% Draw an image.
q
480 0 0 480 50 250 cm
/I0 Do
Q
.sp
% Draw a triangle.
q
1 0 0 rg
50 50 m
100 200 l
200 50 l
f
Q
.sp
% Show some text.
q
0 0 1 rg
BT /TmRm 24 Tf 50 760 Td (Hello, world!) Tj ET
BT /Helv-C 24 Tf 50 730 Td <fac4d2c1d7d3d4d7d5cad4c521> Tj ET
BT /Helv-G 24 Tf 50 700 Td <eae1ebe7ecddf1e1> Tj ET
BT /Song 24 Tf 50 670 Td <4F60 597D> Tj ET
BT /Mincho 24 Tf 50 640 Td <3053 3093 306b 3061 306f> Tj ET
BT /Batang 24 Tf 50 610 Td <c548 b155 d558 c138 c694> Tj ET
Q
.EE
.RE
.
.SH DRAW
.SY mutool
.B draw
.RI [ options ]
.I file
.RI [ pages\~ .\|.\|.]
.YS
.PP
The draw command will render a document to image files,
convert to another vector format, or extract the text content.
.br
The supported input document formats are:
.BR pdf ,\~ xps ,\~ cbz ,\~ md ,
.RB and\~ epub .
.br
The supported output image formats are:
.BR pbm ,\~ pgm ,\~ ppm ,\~ pam ,\~ png ,\~ pwg ,\~ pcl ,
.RB and\~ ps .
.br
The supported output vector formats are:
.BR svg , pdf ,
and
.B debug trace
.RB (as\~ xml ).
.br
The supported output text formats are:
.B plain text
.RB , html ,
and structured text (as
.BR xml\~ or\~ json ).
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.BI \-o\~ output
The output format is inferred from the output filename.
Embed %d in the name to indicate the page number (for example: "page%d.png").
Printf modifiers are supported, for example "%03d".
If no output is specified, the output will go to stdout.
.TP
.BI \-F\~ format
Enforce a specific output format. Only necessary when outputting to stdout
since normally the output filename is used to infer the output format.
.TP
.B \-q
Be quiet, do not print progress messages.
.TP
.BI \-R\~ angle
Rotate clockwise by given number of degrees.
.TP
.BI \-r\~ resolution
Render the page at the specified resolution.
The default resolution is 72 dpi.
.TP
.BI \-w\~ width
Render the page at the specified width (or, if the
.B -r
flag is used, render with a maximum width).
.TP
.BI \-h\~ height
Render the page at the specified height (or, if the
.B -r
flag is used, render with a maximum height).
.TP
.B \-f
Fit exactly; ignore the aspect ratio when matching specified width/heights.
.TP
.BI \-b\~ box
Use named page box
.RB ( MediaBox ,
.BR CropBox ,
.BR BleedBox ,
.BR TrimBox ,
or
.BR ArtBox ).
.TP
.BI \-B\~ bandheight
Render in banded mode with each band no taller than the given height. This uses
less memory during rendering. Only compatible with pam, pgm, ppm, pnm and png
output formats. Banded rendering and md5 checksumming may not be used at the
same time.
.TP
.BI \-T\~ threads
Number of threads to use for rendering (banded mode only).
.TP
.BI \-W\~ width
Page width in points for EPUB layout.
.TP
.BI \-H\~ height
Page height in points for EPUB layout.
.TP
.BI \-S\~ size
Font size in points for EPUB layout.
.TP
.BI \-U\~ filename
User CSS stylesheet for EPUB layout.
.TP
.B \-X
Disable document styles for EPUB layout.
.TP
.B \-a
Disable usage of accelerator file.
.TP
.BI \-c\~ colorspace
Render in the specified colorspace.
Supported colorspaces are:
.BR mono ,
.BR gray ,
.BR grayalpha ,
.BR rgb ,
.BR rgbalpha ,
.BR cmyk ,
.BR cmykalpha .
Some abbreviations are allowed:
.BR m ,
.BR g ,
.BR ga ,
.BR rgba ,
.BR cmyka .
The default is chosen based on the output format.
.TP
.BI \-e\~ filename
Proof ICC profile filename to use for rendering.
.TP
.BI \-G\~ gamma
Apply gamma correction.
Some typical values are 0.7 or 1.4 to thin or darken text rendering.
.TP
.B \-I
Invert colors.
.TP
.BI \-s\~ [mft5]
Show various bits of information:
.B m
for glyph cache and total memory usage,
.B f
for page features such as whether the page is grayscale or color,
.B t
for per page rendering times as well statistics, and
.B 5
for md5 checksums of rendered images that can be used to check if rendering has
changed.
.TP
.BI \-A\~ bits
Specify how many bits of anti-aliasing to use. The default is 8.
.TP
.BI \-A\~ bits/bits
Specify how many bits of anti-aliasing to use. The default is 8. Bits for
graphics is before the slash, bits for text after the slash.
.TP
.BI \-l\~ width
Minimum stroked line width (in pixels).
.TP
.B \-K
Do not draw text.
.TP
.B \-KK
Only draw text.
.TP
.B \-D
Disable use of display lists. May cause slowdowns, but should reduce
the amount of memory used.
.TP
.B \-i
Ignore errors.
.TP
.BI \-m\~ limit
Limit memory usage in bytes.
.TP
.B \-L
Low memory mode (avoid caching objects by clearing cache after each page).
.TP
.B \-P
Run interpretation and rendering at the same time.
.TP
.B \-N
Disable ICC workflow.
.TP
.BI \-M\~ overprint
Control spot/overprint rendering:
.B 0
for no spot rendering,
.B 1
for Overprint simulation Overprint simulation (default), or
.B 2
for full spot rendering.
.TP
.BI \-t\~ language
Specify language/script for OCR (default: eng)
.TP
.BI \-d\~ ocr-file-path
Specify path for OCR files (default: rely on
.B TESSDATA_PREFIX
environment variable.
.TP
.BI \-k\~ correction
Set the skew correction, either one of
.BR auto ,
.B 0
for increase size,
.B 1
for maintain size, or
.B 2
for decrease size.
.TP
.BI \-k\~ correction,angle
Set the skew correction as well as the angle.
.TP
.BI \-y\~ l
Print the layer configs to stderr.
.TP
.BI \-y\~ layer-number
Select layer config (by number from
.B -y l
).
.TP
.BI \-y\~ layer-number\c
.RI , item1\c
.RI , item2\c
, ...
Select layer config (by number from
.B -y l
) and toggle the listed items.
.TP
.B \-Y
Print the individual layers to stderr.
.TP
.BI \-z\~ layer-number
Hide individual layer.
.TP
.BI \-Z\~ layer-number
Show individual layer.
.TP
.I pages
Comma-separated list of page numbers and ranges (for example: 1,5,10-15,20-N),
where the character N denotes the last page. If no pages are specified, then all
pages will be rendered.
.
.SH EXTRACT
.SY mutool
.B extract
.RI [ options ]
.I file.pdf
.RI [ object-number\~ .\|.\|.]
.YS
.PP
The extract command can be used to extract images and font files from a PDF.
If no object numbers are given on the command line, all images and fonts
will be extracted.
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.B \-r
Convert images to RGB when extracting them.
.TP
.B \-a
Embed SMasks as alpha channel..
.TP
.B \-N
Do not use ICC color conversions.
.
.SH INFO
.SY mutool
.B info
.RI [ options ]
.I file.pdf
.RI [ pages\~ .\|.\|.]
.YS
.PP
The info command lists the resources used on each page in a PDF file.
The default is to list all resource types, but if one
or more flags are given, only the flagged types will be shown.
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.B \-F
List fonts.
.TP
.B \-I
List images.
.TP
.B \-M
List page dimensions.
.TP
.B \-P
List patterns.
.TP
.B \-S
List shadings.
.TP
.B \-X
List form and postscript XObjects.
.TP
.B \-Z
List ZUGFeRD info.
.TP
.I pages
Comma-separated list of page numbers and ranges (for example: 1,5,10-15,20-N),
where the character N denotes the last page. If no pages are specified, then all
pages will be included.
.
.SH MERGE
.SY mutool
.B merge
.RI [ options ]
.IR file1\~ [ pages1 ]
.RI [ file2\~ [ pages2 ]
.RI .\|.\|.]
.YS
.PP
The merge command is used to pick out pages from two or more files and merge
them in order into a new PDF output file.
.TP
.BI \-o\~ output
The output filename.
.TP
.BI \-O\~ options
Comma-separated list of format specific output options, see
.B PDF OUTPUT OPTIONS
below.
.
.SH PAGES
.SY mutool
.B pages
.RI [ options ]
.IR input.pdf\~ [ pages\~ .\|.\|.]
.YS
.PP
The pages command dumps information about the size and orientation
of pages within the document.
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.I pages
Comma-separated list of page numbers and ranges (for example: 1,5,10-15,20-N),
where the character N denotes the last page. If no pages are specified, then all
pages will be included.
.
.SH POSTER
.SY mutool
.B poster
.RI [ options ]
.I input.pdf
.RI [ output.pdf ]
.YS
.PP
The poster command splits each page into tiles, and puts each tile on
a page of its own. It's useful for printing a large page onto smaller
pieces of paper that can then be glued together to create a large poster.
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.BI \-m\~ margin
Margin (overlap) between pages in points or percent.
.TP
.BI \-x\~ factor
Split the page into this many horizontal pieces.
.TP
.BI \-y\~ factor
Split the page into this many vertical pieces.
.TP
.B \-r
Splits hoirzontally from right to left (default splits from left to right).
.PP
The output will have x times y number of pages for each input page.
.
.SH RECOLOR
.SY mutool
.B recolor
.RI [ options ]
.I input.pdf
.YS
.PP
Change the color space used in a PDF to a desired one. This command can,
for example, change a color PDF to be grayscale.
.TP
.BI \-c\~ colorspace
Desired output colorspace, choose one of
.BR gray\~ (default),
.BR rgb ,
or
.BR cmyk .
.TP
.BI \-o\~ filename
The output filename.
.TP
.B \-r
Remove any output intents.
.
.SH RUN
.SY mutool
.B run
.I script.js
.RI [ arguments ]
.YS
.PP
Executes a Javascript program which has access to most of the features of the
MuPDF library. The command supports ECMAScript 5 syntax in strict mode. All of
the MuPDF constructors and function live in the global object, and the command
line arguments are accessible from the global argv object.
.PP
If invoke without any arguments, it will drop you into an interactive REPL
(read-eval-print-loop). On the interactive prompt, if you prefix a line with an
equal character it will automatically print the results of the line.
.PP
See the MuPDF documentation at https://mupdf.readthedocs.io for details about
the Javascript interfaces.
.
.SH SHOW
.SY mutool
.B show
.RI [ options ]
.I file.pdf
.RI [ object-number\~ .\|.\|.]
.YS
.PP
The show command will print the specified objects and streams to stdout.
Streams are decoded and non-printable characters are represented
with a period by default.
.TP
.BI \-p\~ password
Use the specified password if the file is encrypted.
.TP
.BI \-o\~ file
Write output to file instead of stdout.
.TP
.B \-e
Print streams in their original encoded (or compressed) form.
.TP
.B \-b
Print streams as binary data and omit the object header.
.TP
.B \-g
Print each object on a single line, suitable for grep.
.TP
.B \-r
Force repair before showing any objects.
.TP
.B \-L
Show object labels (how to reach an object from the root).
.PP

CHANGES
List of changes in MuPDF 1.28

	Markdown document format support!

	New internal options parsing API:
		- Removed fz_has_option, fz_copy_option, and fz_option_eq.
		- Added fz_options struct with constructors and accessors.
		- Detect and warn for unknown and mis-spelled options.
		- The mutool draw command now accepts device specific options with -O.
		- Setting overprint simulation in mutool draw is now done with -M.

	HTML and EPUB improvements:
		- vertical-align for inline images
		- vertical-align percentage based values
		- support tags <u>, <s>, <mark>
		- background-color on inline elements
		- :before and :after pseudo-selectors with content
		- input form elements now have an appearance (but non-interactive)
		- respect meta viewport to set minimum page size

	Per-document control of publisher and user CSS.
		- context global fz_set_use_document_css and fz_set_user_css deprecated
		- replaced with per-document fz_style_document

	PDF write option "decrypt" is deprecated; use "encrypt=none" instead.

	New mutool audit command to analyze PDF structure and generate report

	Many improvements and fixes to incremental search API.
		- customizable structured text to plain text flattening conversions
		- mutool grep command uses regular expressions by default

	Improved text extraction:
		- improved ActualText replacement handling for non-text elements
		- more robust behavior around ActualText replacements
		- presentation form (ligature) decomposition for all scripts
		- improved text-indent paragraph detection
		- improved fake bold detection
		- improved handling of soft and hard hyphens

	Added fz_check_document to scan document and fix possible structural
	errors. Use this to avoid the risk of encountering file errors that may
	cause edits to be lost when the file structure is repaired.

	Support linear and radial shadings when converting XPS (and other formats) to PDF.

	Checked arithmetic functions like in C23 (fz_ckd_mul_int etc)

	Many bug fixes, optimizations, security hardening fixes, and undefined
	behavior code cleanups.

	Building bindings now requires the pipcl package to be installed in the python environment, e.g.:
		python -m venv venv
		source venv/bin/activate
		pip install pipcl
		make c++ ; make python ; make csharp

List of changes in MuPDF 1.27

	New and updated documentation! Build local HTML and Markdown versions from source with sphinx.

	New (experimental) search API that supports matching results across pages and regular expressions.

	Build system:
		Optimize thirdparty dependency submodules to reduce size of git clone and release tarballs.
		Allow conditional compilation for html, xps, svg, extract modules.
		Visual Studio project added an arm64 configuration.
		WASM bindings folded back into main git repository.
		Makefile added new "install-tools" target.

	General:
		More robust document type recognition.
		Many new classes and functions to both WASM and mutool run Javascript bindings.
		Changes to mutool run to bring it into sync with WASM library -- if you stick to ES5 syntax, scripts can run in both environments.
		Improve dehyphenation behavior in structured text extraction.
		Several optimizations when encountering JPEG2000 images in PDF files.
		Fixed many bugs with OCG handling (more improvements to come in a future release).

	HTML and EPUB:
		Hyphenation!
			Set HTML lang attribute and CSS style hyphens:auto to enable.
		CSS non-static positioning:
			position: absolute and relative
			inset/top/right/bottom/left
		HTML tables:
			colgroup and col
			rowspan and colspan
			width, height, valign, and border-collapse
		Legacy HTML attributes:
			width, height, valign, align, bgcolor, border, hidden
		CSS border-style: double, dotted, dashed, outset, inset, ridge, groove
		CSS tree-structural pseudo-selectors (:root, :empty, :first-child, :nth-child, etc)
		CSS :link pseudo-class
		Improved BiDi support with direction:rtl CSS property and dir=rtl HTML attribute
		Recognize all CSS named colors!
		Improved fallback font detection.
		EPUB 3 table of contents.

	PDF Forms & Annotations:
		Respect AcroForm/NeedAppearances flag -- widgets may look different if this flag is set!
		Changed Text (aka Note) annotation sizing and positioning to conform better with the PDF reference.
		Respect NoZoom and NoRotate flags.
		Fixed bug with some checkboxes not toggling as they should.

	Tools & Programs:
		mutool:
			Handle output to stdout more reliably in all tools.
		mutool run:
			allow import statements for mupdf and fs modules
			script compatibility tweaks with node
			process.argv
			process.exit
		mutool clean:
			vectorize text (turn text objects into paths) with -v flag
		mutool draw:
			behavior tweaked to be more consistent across output formats
		mutool poster:
			improve behavior when pages have been rotated
		mutool merge:
			preserve external (only) links when merging documents
		mutool extract:
			extract embedded files and attachments
		mutool grep:
			new tool to search for text in documents
		mupdf-gl:
			confirm before opening potentially dangerous external links.
		mupdf-x11:
			refuse to open potentially dangerous external links.

	API changes:
		Use fz_quad in fz_warp_pixmap, fz_autowarp_pixmap, and fz_detect_document.

	API news:
		- special redaction mode PDF_TEXT_REMOVE_INVISIBLE to strip only invisible text.
		- pdf_check_document (PDFDocument.check in JS and Java) function to scan the file for errors.
		- pdf_set_page_tree_cache to allow faster first-page loads at the cost of slower loading of all other pages.
		- fz_stext_page_details -- stext can now represent text from multiple pages.

List of changes in MuPDF 1.26

	Removed PDF linearization support:

		See https://artifex.com/blog/mupdf-removes-linearisation for
		more information.

	Object labels:

		Write a comment before numbered objects listing the possible
		ways it can be reached from the document root.

		Enable with mutool show -L, mutool clean -L, and PDF write

source/tools/mutool.c
// Copyright (C) 2004-2025 Artifex Software, Inc.
//
// This file is part of MuPDF.
//
// MuPDF is free software: you can redistribute it and/or modify it under the
// terms of the GNU Affero General Public License as published by the Free
// Software Foundation, either version 3 of the License, or (at your option)
// any later version.
//
// MuPDF is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
// details.
//
// You should have received a copy of the GNU Affero General Public License
// along with MuPDF. If not, see <https://www.gnu.org/licenses/agpl-3.0.en.html>
//
// Alternative licensing terms are available from the licensor.
// For commercial licensing, see <https://www.artifex.com/> or contact
// Artifex Software, Inc., 39 Mesa Street, Suite 108A, San Francisco,
// CA 94129, USA, for further information.

/*
 * mutool -- swiss army knife of pdf manipulation tools
 */

#include "mupdf/fitz.h"

#include <string.h>
#include <stdio.h>

#ifdef _MSC_VER
#define main main_utf8
#endif

int muconvert_main(int argc, char *argv[]);
int mudraw_main(int argc, char *argv[]);
int mutrace_main(int argc, char *argv[]);
int murun_main(int argc, char *argv[]);

int pdfclean_main(int argc, char *argv[]);
int pdfextract_main(int argc, char *argv[]);
int pdfinfo_main(int argc, char *argv[]);
int pdfposter_main(int argc, char *argv[]);
int pdfshow_main(int argc, char *argv[]);
int pdfpages_main(int argc, char *argv[]);
int pdfcreate_main(int argc, char *argv[]);
int pdfmerge_main(int argc, char *argv[]);
int pdfsign_main(int argc, char *argv[]);
int pdfrecolor_main(int argc, char *argv[]);
int pdftrim_main(int argc, char *argv[]);
int pdfbake_main(int argc, char *argv[]);
int mubar_main(int argc, char *argv[]);

int cmapdump_main(int argc, char *argv[]);
int pdfaudit_main(int argc, char *argv[]);

static struct {
	int (*func)(int argc, char *argv[]);
	char *name;
	char *desc;
} tools[] = {
#if FZ_ENABLE_JS
	{ murun_main, "run", "run javascript" },
#endif
	{ mudraw_main, "draw", "convert document" },
	{ muconvert_main, "convert", "convert document (with simpler options)" },
#if FZ_ENABLE_PDF
	{ pdfaudit_main, "audit", "produce usage stats from PDF files" },
	{ pdfbake_main, "bake", "bake PDF form into static content" },
	{ pdfclean_main, "clean", "rewrite PDF file" },
	{ pdfcreate_main, "create", "create PDF document" },
	{ pdfextract_main, "extract", "extract font and image resources" },
	{ pdfinfo_main, "info", "show information about PDF resources" },
	{ pdfmerge_main, "merge", "merge pages from multiple PDF sources into a new PDF" },
	{ pdfpages_main, "pages", "show information about PDF pages" },
	{ pdfposter_main, "poster", "split large PDF page into many tiles" },
	{ pdfrecolor_main, "recolor", "change colorspace of PDF document" },
	{ pdfshow_main, "show", "show internal PDF objects" },
	{ pdfsign_main, "sign", "manipulate PDF digital signatures" },
	{ pdftrim_main, "trim", "trim PDF page contents" },
#ifndef NDEBUG
	{ cmapdump_main, "cmapdump", "dump CMap resource as C source file" },
#endif
#endif
	{ mutrace_main, "trace", "trace device calls" },
#if FZ_ENABLE_BARCODE
	{ mubar_main, "barcode", "encode/decode barcodes" },
#endif
};

static int
namematch(const char *end, const char *start, const char *match)
{
	size_t len = strlen(match);
	return ((end-len >= start) && (strncmp(end-len, match, len) == 0));
}

#ifdef GPERF
#include "gperftools/profiler.h"

static int profiled_main(int argc, char **argv);

int main(int argc, char **argv)
{
	int ret;
	ProfilerStart("mutool.prof");
	ret = profiled_main(argc, argv);
	ProfilerStop();
	return ret;
}

static int profiled_main(int argc, char **argv)
#else
int main(int argc, char **argv)
#endif
{
	char *start, *end;
	char buf[32];
	int i;

	if (argc == 0)
	{
		fprintf(stderr, "No command name found!\n");
		return 1;
	}

	/* Check argv[0] */

	if (argc > 0)
	{
		end = start = argv[0];
		while (*end)
			end++;
		if ((end-4 >= start) && (end[-4] == '.') && (end[-3] == 'e') && (end[-2] == 'x') && (end[-1] == 'e'))
			end = end-4;
		for (i = 0; i < (int)nelem(tools); i++)
		{
			strcpy(buf, "mupdf");
			strcat(buf, tools[i].name);
			if (namematch(end, start, buf) || namematch(end, start, buf+2))
				return tools[i].func(argc, argv);
			strcpy(buf, "mu");
			strcat(buf, tools[i].name);
			if (namematch(end, start, buf))
				return tools[i].func(argc, argv);
		}
	}

	/* Check argv[1] */

	if (argc > 1)
	{
		for (i = 0; i < (int)nelem(tools); i++)
			if (!strcmp(tools[i].name, argv[1]))
				return tools[i].func(argc - 1, argv + 1);
		if (!strcmp(argv[1], "-v"))
		{
			fprintf(stderr, "mutool version %s\n", FZ_VERSION);
			return 0;
		}
	}

	/* Print usage */

	fprintf(stderr, "mutool version %s\n", FZ_VERSION);
	fprintf(stderr, "usage: mutool <command> [options]\n");

	for (i = 0; i < (int)nelem(tools); i++)
		fprintf(stderr, "\t%s\t-- %s\n", tools[i].name, tools[i].desc);

	return 1;
}

#ifdef _MSC_VER
int wmain(int argc, wchar_t *wargv[])
{
	char **argv = fz_argv_from_wargv(argc, wargv);
	int ret = main(argc, argv);
	fz_free_argv(argc, argv);
	return ret;
}
#endif
